### PR TITLE
feat: apkam-aware activate_cli

### PR DESCRIPTION
**Pull request by gkc**
**- What I did**
- feat: 'activate' CLI is now APKAM-aware, and supports 
  - onboarding (as before)
  - submitting enrollment requests
  - listing / approving / denying / revoking enrollment requests
  - generating one-time passcodes
  - setting semi-permanent passcode

**- How I did it**
- New code in src/cli : auth_cli.dart, auth_cli_args.dart
- Modified existing bin/activate_cli.dart so it uses auth_cli.main instead of activate_cli.main

**- How to verify it**
- See NoPorts e2e tests ([initial PR](https://github.com/atsign-foundation/noports/pull/1037)) which use this CLI extensively
- Some examples if you wish to test manually:
  - `dart bin/activate_cli.dart -h` to see overall usage info and list of available commands
  - `dart bin/activate_cli.dart <some command> -h` to see help and usage info for a specific command
  - `dart bin/activate_cli.dart otp -a @alice` to generate an OTP for example
  - `dart bin/activate_cli.dart interactive -a @alice` to run interactively

**-Additional context**
- TODO but will be in follow-up PR as it's not on critical path right now: Add code which allows user to 'resume' an enrollment similar to what has been done in at_client_mobile - basically if the CLI is killed while waiting for its enrollment request to be approved or denied, then allow the exact same command to be re-run, but instead of submitting a new enrollment request it finds the details of the previous one and then goes straight to attempting PKAM auth

### CODE REVIEW

File: packages/at_onboarding_cli/lib/src/onboard/at_onboarding_service_impl.dart [hunk 1]
```diff
@@ -131,15 +131,21 @@ class AtOnboardingServiceImpl implements AtOnboardingService {
...
+  }) async {
     if (appName == null || deviceName == null) {
       throw AtEnrollmentException(
           'appName and deviceName are mandatory for enrollment');
     }
-    pkamRetryIntervalMins ??= atOnboardingPreference.apkamAuthRetryDurationMins;
-    final Duration retryInterval = Duration(minutes: pkamRetryIntervalMins);
+    retryInterval ??= Duration(
+        minutes: pkamRetryIntervalMins ??
+            atOnboardingPreference.apkamAuthRetryDurationMins);
```
Comment (gkc) [inline packages/at_onboarding_cli/lib/src/onboard/at_onboarding_service_impl.dart]: added retryInterval as a Duration, so calling code has more control

File: packages/at_onboarding_cli/lib/src/onboard/at_onboarding_service_impl.dart [hunk 2]
```diff
@@ -160,11 +166,11 @@ class AtOnboardingServiceImpl implements AtOnboardingService {
     atLookUpImpl.atChops = AtChopsImpl(atChopsKeys);
 
     // Pkam auth will be attempted asynchronously until enrollment is approved/denied
-    _attemptPkamAuthAsync(
+    await _attemptPkamAuthAsync(
```
Comment (gkc) [inline packages/at_onboarding_cli/lib/src/onboard/at_onboarding_service_impl.dart]: added missing await

File: packages/at_onboarding_cli/lib/src/onboard/at_onboarding_service_impl.dart [hunk 3]
```diff
@@ -174,12 +180,13 @@ class AtOnboardingServiceImpl implements AtOnboardingService {
     return enrollmentResponse;
   }
 
-  void _listenToPkamSuccessStream(
+  Future<void> _listenToPkamSuccessStream (
```
Comment (gkc) [inline packages/at_onboarding_cli/lib/src/onboard/at_onboarding_service_impl.dart]: Made _listenToPkamSuccessStream return a Future which is only completed when we've actually processed the success

File: packages/at_onboarding_cli/lib/src/util/print_full_parser_usage.dart [hunk 4]
```diff
@@ -0,0 +1,52 @@
+import 'dart:io';
+
+import 'package:args/args.dart';
+
+extension PrintAllArgParserUsage on ArgParser {
```
Comment (gkc) [inline packages/at_onboarding_cli/lib/src/util/print_full_parser_usage.dart:5]: A little extension to help with pretty printing of the auth_cli's ArgParsers' usage when there are subcommands involved

File: packages/at_onboarding_cli/lib/src/cli/auth_cli.dart [hunk 5]
```diff
@@ -0,0 +1,661 @@
...
+    stderr.writeln('\n'
+        'Use --help or -h flag to show full usage of all commands'
+        '\n');
+    return 1;
+  }
+
+  final first = arguments.first;
+  if (first.startsWith('-') && first != '-h' && first != '--help') {
+    // no command found ... legacy ... insert 'onboard' as the command
+    arguments = ['onboard', ...arguments];
```
Comment (gkc) [inline packages/at_onboarding_cli/lib/src/cli/auth_cli.dart:50]: This is the only tricky bit of logic to be aware of, to allow us to most easily provide backwards compatibility

**Comment by murali-shris** [comment]
@gkc While running enroll using activate_cli, I am getting this
```
murali@Muralidharans-MacBook-Pro at_onboarding_cli % dart bin/activate_cli.dart enroll -s 1P14P1  -r vip.ve.atsign.zone -p buzz -d pixel -n “ "buzz:rw” -a @alice🛠 -k /Users/murali/.atsign/keys/@alice🛠_buzz.atKeys
dquote> 
```

**Comment by murali-shris** [comment]
@gkc 
```
murali@Muralidharans-MacBook-Pro at_onboarding_cli % dart bin/activate_cli.dart enroll -s 5KSNW3  -r vip.ve.atsign.zone -p buzz -d pixel -n “buzz:rw” -a @alice🛠 -k /Users/murali/.atsign/keys/@alice🛠_buzz.atKeys
```
Can we log the enrollment response from server after running the above command. I had to copy the enrollmentId from the server logs and pass it to the approve command.

**Comment by murali-shris** [comment]
With dart 3.3.4, I am getting this warning every time I run activate_cli
```
murali@Muralidharans-MacBook-Pro at_onboarding_cli % dart bin/activate_cli.dart enroll -s 5KSNW3  -r vip.ve.atsign.zone -p buzz -d pixel -n “buzz:rw” -a @alice🛠 -k /Users/murali/.atsign/keys/@alice🛠_buzz.atKeys
../../../../../.pub-cache/hosted/pub.dev/pointycastle-3.9.0/lib/macs/cbc_block_cipher_mac.dart:103:46: Warning: Operand of null-aware operation '!' has type 'Padding' which excludes null.
 - 'Padding' is from 'package:pointycastle/api.dart' ('../../../../../.pub-cache/hosted/pub.dev/pointycastle-3.9.0/lib/api.dart').
    var paddingName = _padding != null ? '/${_padding!.algorithmName}' : '';
                                             ^
../../../../../.pub-cache/hosted/pub.dev/pointycastle-3.9.0/lib/macs/cbc_block_cipher_mac.dart:203:7: Warning: Operand of null-aware operation '!' has type 'Padding' which excludes null.
 - 'Padding' is from 'package:pointycastle/api.dart' ('../../../../../.pub-cache/hosted/pub.dev/pointycastle-3.9.0/lib/api.dart').
      _padding!.addPadding(_buf, _bufOff);
      ^
../../../../../.pub-cache/hosted/pub.dev/pointycastle-3.9.0/lib/signers/ecdsa_signer.dart:94:41: Warning: Operand of null-aware operation '!' has type 'Mac' which excludes null.
 - 'Mac' is from 'package:pointycastle/api.dart' ('../../../../../.pub-cache/hosted/pub.dev/pointycastle-3.9.0/lib/api.dart').
      kCalculator = _RFC6979KCalculator(_kMac!, n, _pvkey!.d!, message);
                                        ^
../../../../../.pub-cache/hosted/pub.dev/pointycastle-3.9.0/lib/signers/ecdsa_signer.dart:166:7: Warning: Operand of null-aware operation '!' has type 'Digest' which excludes null.
 - 'Digest' is from 'package:pointycastle/api.dart' ('../../../../../.pub-cache/hosted/pub.dev/pointycastle-3.9.0/lib/api.dart').
      _digest!.reset();
      ^
../../../../../.pub-cache/hosted/pub.dev/pointycastle-3.9.0/lib/signers/ecdsa_signer.dart:167:14: Warning: Operand of null-aware operation '!' has type 'Digest' which excludes null.
 - 'Digest' is from 'package:pointycastle/api.dart' ('../../../../../.pub-cache/hosted/pub.dev/pointycastle-3.9.0/lib/api.dart').
      return _digest!.process(message);

```

**Comment by gkc** [comment]
> With dart 3.3.4, I am getting this warning every time I run activate_cli
> 
> ```
> murali@Muralidharans-MacBook-Pro at_onboarding_cli % dart bin/activate_cli.dart enroll -s 5KSNW3  -r vip.ve.atsign.zone -p buzz -d pixel -n “buzz:rw” -a @alice🛠 -k /Users/murali/.atsign/keys/@alice🛠_buzz.atKeys
> ../../../../../.pub-cache/hosted/pub.dev/pointycastle-3.9.0/lib/macs/cbc_block_cipher_mac.dart:103:46: Warning: Operand of null-aware operation '!' has type 'Padding' which excludes null.
>  - 'Padding' is from 'package:pointycastle/api.dart' ('../../../../../.pub-cache/hosted/pub.dev/pointycastle-3.9.0/lib/api.dart').
>     var paddingName = _padding != null ? '/${_padding!.algorithmName}' : '';
>                                              ^
> ../../../../../.pub-cache/hosted/pub.dev/pointycastle-3.9.0/lib/macs/cbc_block_cipher_mac.dart:203:7: Warning: Operand of null-aware operation '!' has type 'Padding' which excludes null.
>  - 'Padding' is from 'package:pointycastle/api.dart' ('../../../../../.pub-cache/hosted/pub.dev/pointycastle-3.9.0/lib/api.dart').
>       _padding!.addPadding(_buf, _bufOff);
>       ^
> ../../../../../.pub-cache/hosted/pub.dev/pointycastle-3.9.0/lib/signers/ecdsa_signer.dart:94:41: Warning: Operand of null-aware operation '!' has type 'Mac' which excludes null.
>  - 'Mac' is from 'package:pointycastle/api.dart' ('../../../../../.pub-cache/hosted/pub.dev/pointycastle-3.9.0/lib/api.dart').
>       kCalculator = _RFC6979KCalculator(_kMac!, n, _pvkey!.d!, message);
>                                         ^
> ../../../../../.pub-cache/hosted/pub.dev/pointycastle-3.9.0/lib/signers/ecdsa_signer.dart:166:7: Warning: Operand of null-aware operation '!' has type 'Digest' which excludes null.
>  - 'Digest' is from 'package:pointycastle/api.dart' ('../../../../../.pub-cache/hosted/pub.dev/pointycastle-3.9.0/lib/api.dart').
>       _digest!.reset();
>       ^
> ../../../../../.pub-cache/hosted/pub.dev/pointycastle-3.9.0/lib/signers/ecdsa_signer.dart:167:14: Warning: Operand of null-aware operation '!' has type 'Digest' which excludes null.
>  - 'Digest' is from 'package:pointycastle/api.dart' ('../../../../../.pub-cache/hosted/pub.dev/pointycastle-3.9.0/lib/api.dart').
>       return _digest!.process(message);
> ```

Should not occur if you do `dart pub upgrade` to pull in the latest pointycastle

**Comment by gkc** [comment]
> @gkc
> 
> ```
> murali@Muralidharans-MacBook-Pro at_onboarding_cli % dart bin/activate_cli.dart enroll -s 5KSNW3  -r vip.ve.atsign.zone -p buzz -d pixel -n “buzz:rw” -a @alice🛠 -k /Users/murali/.atsign/keys/@alice🛠_buzz.atKeys
> ```
> 
> Can we log the enrollment response from server after running the above command. I had to copy the enrollmentId from the server logs and pass it to the approve command.

Done in latest commit

**Comment by gkc** [comment]
@murali-shris please can you re-review?

**Review by sitaram-kalluri — APPROVED** [review]

**Pull request merged**